1 add data getters

eglucas opened the pull request:
---

# Description

Fixes #1 

This PR adds code to get data for:
- ASHP installation costs
- Property heat demand
- DESNZ wholesale price projections
- Ofgem energy price cap (as `Tariff` objects from `asf_levies_model`)
- Levy rebalancing results (as rebalanced `Levy` objects from `asf_levies_model`)

The main data getter functions are `getters/data_getters.py`, with some helper functions in `getters/getter_utils.py`.

Note, there are parameters defined in the first section of the `get_levies()` function which are necessary for the rebalancing calculations. I've left these hard-coded for now, but they do need to be updated each time there is new DESNZ consumption data available (annual) and Ofgem price cap data (quarterly). I think how we assign these parameters can be refactored later with their own getters.

This PR also updates the `pyproject.toml` and `uv.lock` files with dependencies. 

# Instructions for Reviewer

For this PR, please can you look at the code in `data_getters.py` and `getter_utils.py`. You can explore their functionality using `data_getters_demo.py` (please open as a notebook), where I have also provided some information on how the `Levy`, `LevyCollection` and `Tariff` objects all slot in together.

I also have documented some assumptions/changes in this[ Google doc](https://docs.google.com/document/d/1hSefJigBTMyI4tRfPJ9yAPsSLkXUMpR2a2qbakTROB0/edit?tab=t.8kmclo13s88).

Please pay special attention to anything that needs more documentation, errors in logic and any improvement opportunities.

# Checklist:

- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

eglucas committed:
update dependencies

eglucas committed:
add data getters

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+"""Data getters for inputs into lifetime cost calculations."""
```
Minor, but I'd add a list of the getters here for example something like:

Data getters for inputs into lifetime cost calculations including:
- air source hp installation costs
- gas boiler installation costs
- ...

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+
+    Returns:
+        pd.DataFrame: Dataframe of air-source heat pump installation \
+            costs for different property archetypes.
```
also decile?

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+    Get dataframe of air-source heat pump installations costs from S3.
+
+    Returns:
+        pd.DataFrame: Dataframe of air-source heat pump installation \
```
where these `\` added automatically by `pre-commit`?

eglucas replied:
No, it was me manually placing them - but I did them in a slapdash way so it's not consistent. I agree it doesn't necessarily make the text more readable. I've removed all the line breaks for now and I can go back and alter them after we discuss what convention we want to follow for the rest of the repo!

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+            payment method and price cap period provided.
+    """
+
+    # Get Annex 9
```
Minor, but you could add more info about what's in Annex 9, such as

"Get Annex 9 which includes..."

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+        Tuple[Tariff, Tariff]: Gas Tariff and electricity Tariff objects \
+            corresponding to payment method for current price cap.
+    """
+    current_price_cap_period = datetime.now().strftime("%Y-%m-%d")
```
can this be any date? or does it need to be a date in a specific set?

eglucas replied:
This `get_current_energy_price_cap_tariffs` function uses the generic `get_tariffs` function, and the `price_cap_period` argument's accepted format is described in the docstring for `get_tariffs` :)

> price_cap_period (str): Date of interest in YYYY-MM-DD format. "LATEST" is also valid to get the most recently available price cap.

danlewis85 replied:
This can be any date with the range of dates for which there is a price cap.

This means it cannot be a date in the future beyond the end of the latest price cap for the annex of interest.

It also cannot pre-date price caps.

Within that range, the Tariff or Levy object will return the price cap that contains the date you provide.

sofiapinto replied:
So I'm guessing that doing something like

```
electricity_tariff = tariffs.ElectricityStandardCredit.from_dataframe(
            levies_data_getters.process_tariff_elec_standard_credit_nil(fileobject),
            levies_data_getters.process_tariff_elec_standard_credit_typical(fileobject),
            price_cap=price_cap_period,
        )
```
will yield an error if the `price_cap_period` is not within the possible range of dates?

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+            each representing a levy present in the policy costs component of the price cap period provided.
+    """
+
+    # Denominator values from DESNZ subnational consumption domestic data, 2023
+    supply_elec = 96_517_461
+    supply_gas = 266_505_188
+    customers_gas = 24_605_467
+    customers_elec = 29_239_936
```
What do these values represent exactly? Do we have a link to this data?

sofiapinto replied:
These can definitely go into a config file with the documentation you have here + perhaps the link to the data.

I wonder, how hard is it to read these directly from their source, rather than hardcoding them? Not for now, but for future. I know you said some of these are updated yearly and others quarterly.

eglucas replied:
I've added more documentation that explains each parameter being defined. Have also added the source URLs. My plan is to leave all assignments in these getter functions for now but will migrate to config once we have finalised the structure of the config file(s).

eglucas replied:
And yes, I think it will be relatively easy to read these from source and we can definitely think about creating their own getter functions later.

sofiapinto commented on uv.lock:
i'm guessing this was all added automatically and I don't have to look at any of this :)

sofiapinto commented on pyproject.toml:
```diff
@@ -13,7 +13,15 @@ authors = [{ name = "Nesta" }]
+    "boto3>=1.39.15",
+    "odfpy>=1.4.1",
+    "pandas>=2.3.1",
+    "pandera>=0.25.0",
```
wondering why we have pandera here :o

eglucas replied:
it's a dependency for some of the functions in `asf_levies_model.getters.load_data` and it wouldn't let me import that module until i also had pandera installed! we used it for some initial work on data type validation

sofiapinto commented on pyproject.toml:
```diff
@@ -54,3 +62,6 @@ convention = "google"
 docstring-code-line-length = "dynamic"
+
+[tool.uv.sources]
+asf-levies-model = { git = "https://github.com/nestauk/asf_levies_model.git" }
```
interesting how these are added for uv. was this something you manually added?

eglucas replied:
this was automatically added when i ran `uv add git+https://github.com/nestauk/asf_levies_model.git` after installing it to update pyproject.toml

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+    # Scaling factor for estimating domestic share of NCC revenue
+    ncc_eligible_supply = (
+        119_380_310.7  # Mar-Jun2025 period, Annex 4, NCC methodology tab
+    )
```
same as above about configs/source link etc.

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+    total_supply_elec = (
+        249_044_438  # DESNZ GB total electricity consumption - all meters (2023)
+    )
+    exempt_eii_supply = (
```
what does this stand for?

danlewis85 replied:
energy intensive industry

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+    )
+    fit_scaling_factor = supply_elec / (total_supply_elec - exempt_eii_supply)
+
+    # Scaling factor for estimating domestic share of NCC revenue
```
what's this NCC revenue?

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+                "new_variable_weight_elec": 0,
+                "new_fixed_weight_elec": 0,
+                "new_variable_weight_gas": 0,
+                "new_fixed_weight_gas": 0,
```
i'm guessing these 0's and 1's have a yes/no meaning

eglucas replied:
as we've since discussed, these values are actually proportions. will add some more documentation to make this clearer

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+    ).rebalance_to_denominators()
+
+    rebalancing_weights = create_scenario_weights_dict(levy_collection_for_rebalancing)
+
```
for the scenarios below, is this code copied from somewhere? just wondering if this has been reviewed previously. Although things might make sense to me, I just want to make sure someone who actually understands these has already reviewed this.

danlewis85 replied:
The general process has been reviewed as part of the levies work, however it is also useful to do an 'expert' review of parameters, where typographical errors can occur.

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+                "new_tax_weight": 0,
+                "new_variable_weight_elec": 0,
+                "new_fixed_weight_elec": 0,
+                "new_variable_weight_gas": 1,  # assumes no hybrid of unit cost/standing charge
+                "new_fixed_weight_gas": 0,  # assumes no hybrid of unit cost/standing charge
```
are these supposed to be 1 and 0 and have the same doc?

eglucas replied:
Yes, they are supposed to have the same doc. 

Levies can be charged as unit costs (i.e. variable costs, £/MWh) or as a standing charge (i.e. fixed costs, £/customer/year). What this doc is saying is that when we rebalance from electricity to gas, we are assuming that no levy has a hybrid of standing charge and unit costs, only unit costs. This is true of all current levies (they're either unit costs or standing charges, not a mixture of both).

`new_variable_weight_gas` is the rebalanced proportion of the policy scheme revenue raised by gas bills via unit costs (variable), which we are assuming is 100%. 

As discussed as well, I will ask Dan to review the levy rebalancing scenario definition code.

danlewis85 replied:
Yes, interesting I think this is the right way to do this here because some levies are on both gas and electricity, so you have to specifically set to 1 otherwise you'll lose some levy!

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+        for levy in [
+            levy
+            for levy in levy_collection_for_rebalancing
+            if levy.electricity_weight > 0
```
what happens if levy.electricity_weight =0?

eglucas replied:
With this list comprehension, I only want levies where its `electricity_weight` is NOT zero (i.e. it is currently charged on electricity bills)

danlewis85 replied:
This is fine, however an alternative approach that is more similar to the examples above is:

```{python}
for levy in levy_collection_for_rebalancing[[levy.short_name for levy in levy_collection_for_rebalancing if levy.electricity_weight > 0]]:
    # Subsequent code
```

The advantage of this is that it keeps the Levy Collection rather than a list as the subject of the loop. You can also decouple the two thigns for clarity if you want, e.g.

```{python}
# Get list of short names of levies on electricity.
levies_on_electricity = [levy.short_name for levy in levy_collection_for_rebalancing if levy.electricity_weight > 0]
# Create rebalancing scenario
for levy in levy_collection_for_rebalancing[levies_on_electricity]:
    # Subsequent code
```

danlewis85 replied:
Also, this selection isn't _strictly_ necessary as you should get the same result iterating over all levies regardless and applying the rebalancing weights below. However, being explicit is good practice.

eglucas replied:
Thanks Dan! Good point, I lifted this code out from a file that must've been from before we introduced the LevyCollection object. I'll make that change because it is cleaner.

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+        for levy in [
+            levy
+            for levy in levy_collection_for_rebalancing
+            if levy.electricity_variable_weight > 0
```
what happens if levy. electricity_variable_weight =0?

eglucas replied:
Same as above - here, I only want to rebalance levies that ARE charged on electricity units and exclude ones that are not charged on electricity units (which will have electricity_variable_weight = 0).

danlewis85 replied:
As above, you could refactor this to keep the levy collection as the subject of the loop, but there are no issues with how this performs otherwise.

sofiapinto commented on asf_lifetime_cost_model/getters/getter_utils.py:
Minor again, but I like to see dosctrings at the top of all files. Could you add:
"""Getter utils to read CSV's from S3 and excel files from given URLs."""
or similar

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+    payment_method: str = "Other Payment Method",
+) -> Tuple[Tariff, Tariff]:
+    """
+    Create gas and electricity Tariff objects from Ofgem price cap data.
```
i wonder if we could add a little bit to what these objects are/hold

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+    Rebalance levies according to supplied denominators (consumption and customer charging base). \
+        Note: This is used for internal consistency in rebalancing as different charging base numbers are \
+            used across different levies to determine levy rates.
+    """
```
i find it more readable like this

```suggestion
    # Rebalance levies according to supplied denominators (consumption and customer charging base). \
    # Note: This is used for internal consistency in rebalancing as different charging base numbers are \
    # used across different levies to determine levy rates.
    """
```

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+        price_cap_period=price_cap_period
+    ).rebalance_to_denominators()
+
+    rebalancing_weights = create_scenario_weights_dict(levy_collection_for_rebalancing)
```
what's the output of this one?

sofiapinto commented on asf_lifetime_cost_model/notebooks/data_getters_demo.py:
Did you create the notebooks folder or was it there from the cookiecutter?
in this 'standard' format of the cookiecutter there's a notebooks folder inside the analysis, so we could move this there to avoid multiple notebooks folders.

eglucas replied:
Ah sorry I didn't realise the notebooks folder was now in analysis! I did create a new one, but have now deleted that and moved the demo notebook to the correct folder.

sofiapinto reviewed (commented):
hey @eglucas - thanks so much for this great work! ✨ 

I left comments throughout. Mostly just minor suggestions or questions.

I need to take a second look at this after reading your assumptions in the document. And I still need to run the notebook - have messaged you on Slack about that.

I'll come back with a final review :)

eglucas committed:
move data_getters_demo notebook

eglucas committed:
add top docstring

eglucas committed:
add more documentation

eglucas committed:
modify line breaks

eglucas commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+    return levy_collection
+
+
+def get_rebalanced_levies(
```
@danlewis85 Thank you for agreeing to look at this code - please could you take a look at the levy rebalancing scenarios that are defined in this function? More specifically, please can you check whether the list comprehension and rebalancing weight definition logic makes sense.

eglucas committed:
update scenario names

eglucas committed:
update wholesale price scenario names

danlewis85 commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+    """
+
+    # Get Annex 9 which contains data on final energy price cap rates
+    fileobject = levies_data_getters.download_annex_9(as_fileobject=True)
```
You don't have to use a fileobject here, you could download the annex, which is the default behaviour of this function, and store it in s3 and use that. You have to update the config anyway to move to more recent annex's so it's not like this implicitly ensures you're using the most uptodate data.

It doesn't necessarily matter that much, but by only use the fileobject variant, you're not archiving a copy of the underlying data.

The original use case for the fileobject was the web app - having an file in memory is convenient if you want to avoid hitting physical storage or aren't confident that it will be persisted.

danlewis85 commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+        266_505_188  # total domestic gas consumption in MWh, GB, non-weather corrected
+    )
+    customers_gas = 24_605_467  # number of domestic gas meters, GB
+    customers_elec = 29_239_936  # number of domestic electricity meters, GB
```
Hard coding these in the data getter may be undesirable - suggest you move them either to config and/or expose in function signature.

sofiapinto replied:
Yes - this has been discussed in one the comments above @danlewis85 , it will be moved to the config :)

danlewis85 commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+            scaling_factor=fit_scaling_factor,
+            price_cap=price_cap_period,
+        ),
+        levies.NCC.from_dataframe(
```
How does this respond for price caps that pre-date NCC?

eglucas replied:
Testing it with a price cap period before it was introduced instantiates the NCC levy with zero rates. Interestingly though it has the correct electricity weights but since revenue is nan then there are no rates.

`{'name': 'Network Charging Compensation Scheme',
 'short_name': 'ncc',
 'electricity_weight': 1,
 'gas_weight': 0,
 'tax_weight': 0,
 'electricity_variable_weight': 1,
 'electricity_fixed_weight': 0,
 'gas_variable_weight': 0,
 'gas_fixed_weight': 0,
 'electricity_variable_rate': nan,
 'electricity_fixed_rate': 0,
 'gas_variable_rate': 0,
 'gas_fixed_rate': 0,
 'general_taxation': 0,
 'revenue': nan,
 'price_cap_period': 01/01/2024 - 31/03/2024,
 'UpdateDate': Timestamp('2023-11-01 00:00:00'),
 'SchemeYear': '2023/2024',
 'EstimatedLevyFund': nan,
 'AdminCosts': nan,
 'ReserveFund': nan,
 'EligibleDemand': nan}`

danlewis85 replied:
I think that's fine, dependent on whether it then affects calculations if it is part of a Levy Collection.

eglucas replied:
Really good point - I tested this and when there is an invalid ncc levy, the `LevyCollection.calculate_variable_levies(electricity_consumption=1, gas_consumption=0)` breaks and outputs nan. I've added a conditional block to the function where ncc is removed from the list if it is found to be invalid.

danlewis85 commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+
+    # Scaling factor for estimating domestic share of Network Charging Compensation revenue
+    ncc_eligible_supply = (
+        119_380_310.7  # Mar-Jun2025 period, Annex 4, NCC methodology tab
```
Does this need to change according to price cap period?

eglucas replied:
As above for removing the need to hard-code by reading from the NCC levy object

danlewis85 commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+    # Scaling factor for estimating domestic share of Feed-in Tariff revenue based on total GB electricity supply and exempt supply for Energy Intensive Industries
+    total_supply_elec = 249_044_438  # DESNZ GB total electricity consumption, 2023, all consumption (domestic and non-domestic)
+    exempt_eii_supply = (
+        10_529_633  # Apr-Jun2025 period, Annex 4, New FIT methodology tab
```
Does this need to change with price cap period?

eglucas replied:
Yes, Exempt EII supply and NCC eligible supply are directly from the FIT and NCC tabs of Annex 4 and are actually attributes of the FIT and NCC levy objects.

This is legacy code and I will make changes now so that it assigns these variables from the levy object attributes.

danlewis85 commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+            }
+
+    else:
+        raise KeyError("Unrecognised scenario name.")
```
I think `KeyError` is misleading. I think a better error to raise is `ValueError`.

danlewis85 commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+            levy
+            for levy in levy_collection_for_rebalancing
+            if levy.electricity_weight > 0
+        ]:
```
As above for defining the levies to update.

danlewis85 commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,394 @@
+                "new_tax_weight": 0,
+                "new_variable_weight_elec": 0,
+                "new_fixed_weight_elec": 0,
+                "new_variable_weight_gas": levy.electricity_variable_weight,
```
Check this gives you the answer you want for levies that are split across gas and electricity.

I think you might instead want to do something like:

```{python}
"new_variable_weight_gas": 1 if levy.electricity_variable_weight > 0 else 0,
"new_fixed_weight_gas":  1 if levy.electricity_fixed_weight > 0 else 0,
```
Again, using the fact that no levies are split across unit and standing charge.

danlewis85 replied:
Basically, check ECO4, GBIS and WHD to see if this works as expected.

eglucas replied:
I've just tested my initial logic with ECO4, GBIS and WHD and it does work as expected (as well as the other levies that are originally fully on electricity and are being fully moved to gas).

If a levy is split between gas/electricity, and we know they are not split across unit and standing charge then, e.g. ,gas_variable_weight = 1 and electricity_variable_weight = 1, and gas_fixed_weight = 0 and electricity_fixed_weight = 0. 

Rebalancing to gas in these cases effectively only zero out the electricity weights, and overwrites the gas weights with the same value (i.e. 1). 

I think that both of the approaches we've come up with have the same logic and outcome because electricity_variable_weight and electricity_fixed_weight are only 0 or 1 - i.e. there's no need to do the weight > 0 conditional.

danlewis85 replied:
That makes sense, those criteria are essentially hierarchical, so only dependent on whether electricity or gas is turned on, not on each other:

```
             / - fixed
electricity <
             \ _ variable

     / - fixed
gas <
     \_ variable

taxation -

```
Fine.

eglucas committed:
lowercase scenario names

eglucas committed:
improve rebalanced levies getter

eglucas committed:
update rebalancing scenario name

eglucas committed:
fix historic ncc levy

eglucas commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -303,9 +303,22 @@ def get_levies(price_cap_period: str) -> LevyCollection:
-    levy_collection = levies.LevyCollection(
-        "Policy Costs", "pc", list_levies, denominator_values
-    )
+    # NCC is a levy that was introduced in price cap period Apr2025-Sep2025
+    # If price cap period of interest is before this, the NCC levy introduces nan values
+    # that creates problems for LevyCollection calculations
+    # Check if NCC has nan revenue
+    ncc_levy = [levy for levy in list_levies if levy.short_name == "ncc"][0]
+    if pd.isna(ncc_levy.revenue):
+        list_levies_ncc_removed = [
+            levy for levy in list_levies if levy.short_name != "ncc"
+        ]
+        levy_collection = levies.LevyCollection(
+            "Policy Costs", "pc", list_levies_ncc_removed, denominator_values
+        )
+    else:
+        levy_collection = levies.LevyCollection(
+            "Policy Costs", "pc", list_levies, denominator_values
+        )
```
@danlewis85 this is the ncc check changes I referred to in other thread

eglucas committed:
modify rebalanced levies getter

eglucas committed:
add ncc levy modification demo

eglucas commented:
Hi @sofiapinto,

Based on the levy rebalancing features requested by the policy team, I've [modified the `get_rebalanced_levies()` data getter](https://github.com/nestauk/asf_lifetime_cost_model/pull/2/commits/b61ec5e89419e3ae5a313822cd43a2ce7c1baee0).

Instead of the argument being a name of rebalancing scenario where the corresponding rebalancing weights are coded in the function itself, the rebalancing weights themselves are now function arguments. I have also added another argument to specify which levies are to be rebalanced.

I think this approach will work best with the streamlit app because the variable set by the slider between gas <> electricity (i.e. electricity_weight and gas_weight) will then be the inputs to this function.

The rebalancing options in the app could then be called:
- "rebalance unit costs between electricity and gas" (can be implemented as a slider that sets `electricity_weight` and `gas_weight` for levies that are currently charged as unit costs)
- "remove all electricity levies to taxation" (can be implemented as a checkbox that will set `electricity_weight=0` and `tax_weight=1` for levies that are currently on electricity)

I've also modified the data_getters_demo notebook to show how this new getter works.

Please let me know if anything is unclear!

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+- annual heat demand for property archetype
+- DESNZ gas and electricity wholesale price projections (2023-2050)
+- Ofgem energy price cap
+- policy costs after levy rebalancing"""
```
```suggestion
"""Data getters for inputs into lifetime cost calculations including:
- inflation adjusted air source heat pump installation costs per decile and property archetype
- annual heat demand for each property archetype
- DESNZ gas and electricity wholesale price projections for natural gas and electricity (2023-2050)
- Ofgem energy price cap
- policy costs after levy rebalancing"""
```

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+
+    Returns:
+        pd.DataFrame: Dataframe of air-source heat pump installation costs for each decile in different property archetypes.
+    """
```
```suggestion
    """
    Get dataframe of inflation-adjusted air-source heat pump installations costs for each decile in different property archetypes from S3.

    Returns:
        pd.DataFrame: Dataframe of air-source heat pump installation costs
    """
```

sofiapinto replied:
just moved some of the info up before the return

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+
+    Returns:
+        pd.DataFrame: Dataframe of average heat demand for each decile in different property archetypes. Deciles are based on ashp installation costs.
+    """
```
```suggestion
    """
    Get dataframe of average heat demand data from S3 for each property archetype.

    Returns:
        pd.DataFrame: Dataframe of average heat demand
    """
```

sofiapinto replied:
same as above but I also removed the info about deciles, because the data is not split by deciles

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+) -> pd.DataFrame:
+    """
+    Get dataframe containing DESNZ wholesale price projections for natural gas and electricity from 2001 to 2050 under different scenarios.
+    Prices are provided in 2023 prices.
```
```suggestion
    Downloads DESNZ wholesale price projections and creates dataframe containing price projections for natural gas and electricity from 2001 to 2050 (inflation-adjusted to 2023 prices) under different scenarios.
    
    The resulting DataFrame contains one column per year (with price data) and columns containing additional information such as: metric, fuel, units and projection_scenario
```

sofiapinto replied:
Is this correct or did I totally misinterpret this?

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+
+        # Filter for rows of interest
+        df = df[
+            (df["fuel"] == "Electricity (volume weighted)")
```
I opened the file and I was wondering what the difference between _volume weighted_ and _baseload_ is, for electricity, and why did we choose volume weighted?

eglucas replied:
This is something we need to confirm with Martina!

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+
+
+def get_desnz_wholesale_price_projections(
+    projection_scenario_names: list[str],
```
why did you decide to have this as an parameter? Why not just read all of the different scenarios as a default?

eglucas replied:
i think at the time i thought it was more efficient to only return scenarios of interest - but i do think it makes more sense to read all of it. will change this!

eglucas replied:
Note that I will need to make changes in the running costs function code as a result of this

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+    """
+
+    # Read all sheets in Excel workbook into a dictionary from website
+    all_sheets = _read_excel_to_dataframe(
```
```suggestion
    all_sheets = _read_excel_to_dataframe_or_dict(
```

sofiapinto commented on asf_lifetime_cost_model/getters/getter_utils.py:
```diff
@@ -0,0 +1,53 @@
+    return pd.read_csv(content)
+
+
+def _read_excel_to_dataframe(file_url: str) -> Optional[dict[str, pd.DataFrame]]:
```
```suggestion
def _read_excel_to_dataframe_or_dict(file_url: str) -> Optional[dict[str, pd.DataFrame]]:
```

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+        df = df.drop(["coverage", "note"], axis=1)
+
+        # Look up scenario name to add to dataframe
+        df["projection scenario"] = scenario
```
```suggestion
        df["projection_scenario"] = scenario
```

eglucas replied:
Was wondering why it would be preferential to have an underscore in the dataframe column name? the rest of the columns in this dataframe don't have underscores

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+    Get dataframe containing DESNZ wholesale price projections for natural gas and electricity from 2001 to 2050 under different scenarios.
+    Prices are provided in 2023 prices.
+
+    Possible scenarios include:
```
```suggestion
    Scenarios for which data is available include:
```

sofiapinto replied:
in case you decide to remove the `projection_scenario_names` parameter

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+    A Tariff object is a representation of the rates that are charged against energy consumption.
+    Attributes include fuel type and price cap period interval, and also hold values of each
+    cost component that contributes to the total cost of energy as a standing charge and as a unit cost.
+    See full documentation at: https://github.com/nestauk/asf_levies_model/blob/dev/asf_levies_model/tariffs.py
```
```suggestion
    Downloads Ofgem price cap data (Annex 9 file) and creates gas and electricity Tariff objects.
    
    A Tariff object is a representation of the rates that are charged against energy consumption.
    Each tariff object includes attributes such as fuel type and price cap period interval, also holding values of each
    cost component that contributes to the total cost of energy as a standing charge and as a unit cost.
    See full documentation at: https://github.com/nestauk/asf_levies_model/blob/dev/asf_levies_model/tariffs.py
```

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+    return combined_projection_scenarios
+
+
+def get_tariffs(payment_method: str, price_cap_period: str) -> Tuple[Tariff, Tariff]:
```
```suggestion
def _create_tariff_objects(payment_method: str, price_cap_period: str) -> Tuple[Tariff, Tariff]:
```

sofiapinto replied:
would this make sense? disregard if not

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+    """
+
+    # Get Annex 9 which contains data on final energy price cap rates
+    fileobject = levies_data_getters.download_annex_9(as_fileobject=True)
```
I do see Dan's point above, and might be a good thing to do in a separate PR. Downloading the data directly to S3 and then reading from there so that, even if the link breaks, we always have the data.

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+    See full documentation at: https://github.com/nestauk/asf_levies_model/blob/dev/asf_levies_model/tariffs.py
+
+    Args:
+        payment_method (str): Payment method of interest, valid arguments are: Other Payment Method, PPM, Standard Credit.
```
```suggestion
        payment_method (str): Payment method of interest, valid arguments are: 'Other Payment Method', 'PPM', 'Standard Credit'.
```

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+    Create gas and electricity Tariff objects from Ofgem price cap data.
+
+    Args:
+        payment_method (str, optional): Payment method of interest, valid arguments are: Other Payment Method, PPM, Standard Credit. Defaults to "Other Payment Method".
```
```suggestion
        payment_method (str, optional): Payment method of interest, valid arguments are: 'Other Payment Method', 'PPM', 'Standard Credit'. Defaults to "Other Payment Method".
```

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+
+    gas_tariff, electricity_tariff = get_tariffs(
+        payment_method=payment_method, price_cap_period=current_price_cap_period
+    )
```
```suggestion
    gas_tariff, electricity_tariff = _create_tariff_objects(
        payment_method=payment_method, price_cap_period=current_price_cap_period
    )
```

sofiapinto replied:
if you decide to change the function name

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+
+    # Definining parameters that are required for levy rebalancing calculations
+
+    # Total domestic energy consumption and energy customer numbers from DESNZ subnational consumption domestic data, 2023 - TO MOVE TO CONFIG
+    # These are to provide consistent charging bases across all levies when rebalancing
+    # Source: https://www.gov.uk/government/statistics/regional-and-local-authority-gas-consumption-statistics
+    # Source: https://www.gov.uk/government/statistics/regional-and-local-authority-electricity-consumption-statistics
+    domestic_supply_electricity = (
+        96_517_461  # total domestic electricity consumption in MWh, GB
+    )
+    domestic_supply_gas = (
+        266_505_188  # total domestic gas consumption in MWh, GB, non-weather corrected
+    )
+    domestic_customers_gas = 24_605_467  # number of domestic gas meters, GB
+    domestic_customers_electricity = (
+        29_239_936  # number of domestic electricity meters, GB
+    )
+    total_supply_electricity = 249_044_438  # DESNZ GB total electricity consumption, 2023, all consumption (domestic and non-domestic)
```
Thank you for adding the data sources @eglucas

I was trying to find these numbers, but can't quite find the same ones? How did you get these? When we open the excel files and go to the 2023 tab, are these obtained by summing the values in those columns? (I'm guessing you just copied the values from Martina's spreadsheet for now)

I saw an answer from you in a separate comment where you said how it would be relatively easy to read these from the source directly/have their own data getter. If these are indeed the results of reading from a specific source and summing a few columns, than it would be a great idea to move in that direction. As we know, optimising all of the getters was not a priority for now, so this is a task for the future.

eglucas replied:
The numbers are lifted straight out of the gov.uk Excel files (no summing) -

For the gas values, from https://www.gov.uk/government/statistics/regional-and-local-authority-gas-consumption-statistics:
- File: Subnational gas consumption statistics, non-weather-corrected 2015 to 2023
- Tab: 2023
- Domestic consumption (which we are also treating as supply) is under the column "Total consumption (GWh): Domestic: cell H7
- Domestic customers is under the column "Number of meters (thousands): Domestic": cell E7

For the electricity values, from https://www.gov.uk/government/statistics/regional-and-local-authority-electricity-consumption-statistics:
- File: Subnational electricity consumption statistics 2005 to 2023
- Tab: 2023
- Domestic consumption: cell K6
- Domestic meters: cell F6

These aren't in Martina's spreadsheet because these are parameters used for levy rebalancing calculations. Martina's spreadsheet only has hard-coded values for levy rates _after_ rebalancing.

Yes, I definitely want to read these from source directly in future with their own getters!

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+    # Get Ofgem Annex 4 Policy Cost model file
+    fileobject = levies_data_getters.download_annex_4(as_fileobject=True)
+
+    # Calculate scaling factor for estimating domestic share of Feed-in Tariff revenue based on total GB electricity supply and exempt supply for Energy Intensive Industries
```
```suggestion
    # Calculate scaling factor for estimating domestic share of Feed-in Tariff (FIT) revenue based on total GB electricity supply and exempt supply for Energy Intensive Industries (EII)
```

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+        total_supply_electricity - fit_exempt_eii_supply
+    )
+
+    # Calculate scaling factor for estimating domestic share of Network Charging Compensation revenue
```
```suggestion
    # Calculate scaling factor for estimating domestic share of Network Charging Compensation (NCC) revenue
```

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+    else:
+        levy_collection = levies.LevyCollection(
+            "Policy Costs", "pc", list_levies, denominator_values
+        )
```
To reduce the amount of code, I think you can remove the code where you add the ncc_levy to the list `list_levies` (when you define the list) and then just do:

```
if not pd.isna(ncc_levy.revenue):
        list_levies.append(levies.NCC.from_dataframe(levies_data_getters.process_data_NCC(fileobject),
            scaling_factor=ncc_scaling_factor,
            price_cap=price_cap_period,
        )

fileobject.close()

levy_collection = levies.LevyCollection(
            name = "Policy Costs",  short_name = "pc", levies = list_levies, denominators = denominator_values
        )
```

eglucas replied:
thanks! will integrate.

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -0,0 +1,382 @@
+) -> LevyCollection:
+    """
+    Creates a LevyCollection where a given set of levies are rebalanced using the same rebalancing weights.
+
+    Args:
+        levies_to_rebalance (list[str]): list containing short names of levies to be rebalanced with provided weights.
+        electricity_weight (float): [0, 1] indicating electricity proportion of levy revenue.
+        gas_weight (float): [0, 1] indicating electricity proportion of levy revenue.
+        tax_weight (float): [0, 1] indicating electricity proportion of levy revenue.
+        variable_electricity_weight (float): [0, 1] indicating the proportion of electricity revenue that is variable (e.g. per unit consumption).
+        fixed_electricity_weight (float): [0, 1] indicating the proportion of electricity revenue that is fixed (e.g. per customer).
+        variable_gas_weight (float): [0, 1] indicating the proportion of gas revenue that is variable (e.g. per unit consumption).
+        fixed_gas_weight (float): [0, 1] indicating the proportion of gas revenue that is fixed (e.g. per customer).
+        price_cap_period (str, optional): Date of interest in YYYY-MM-DD format. Defaults to "LATEST" to get the most recently available price cap.
```
Would it make sense to add a small note explaining what 0 and 1 mean, in each case or would that interpretation be obvious to anyone?

eglucas replied:
I'll add further explanation.

sofiapinto commented on asf_lifetime_cost_model/analysis/notebooks/data_getters_demo.py:
```diff
@@ -0,0 +1,215 @@
+# ### Demonstrating how the data getters work
+
+# %%
+import pandas as pd
```
can be removed, not used :)

sofiapinto commented on asf_lifetime_cost_model/analysis/notebooks/data_getters_demo.py:
```diff
@@ -0,0 +1,215 @@
+# - Units are £ (2023 prices)
+
+# %%
+data_getters.get_ashp_installation_costs()
+
+# %% [markdown]
+# **Property heat demand**
+# - Read from .csv in s3
+# - Units are kWh/year
+
+# %%
+data_getters.get_property_heat_demand()
```
a note for future reference (I'll add it to a separate issue): the archetype labels are different across these two sources

eglucas replied:
Yes, the full label names are different but the archetype numbers are the same! for the running costs function, i've implemented the archetype argument to be selecting the archetype by number not full label name. I do agree though that we should standardise the full label names in future.

sofiapinto commented on asf_lifetime_cost_model/analysis/notebooks/data_getters_demo.py:
```diff
@@ -0,0 +1,215 @@
+current_gas_tariff.__dict__
+
+# %% [markdown]
+# If we want to extract the unit cost of gas:
+
+# %%
+current_gas_tariff.calculate_variable_consumption(1)  # in £/MWh
```
```suggestion
# If we want to extract the unit cost of gas (consumption of 1 unit):

# %%
current_gas_tariff.calculate_variable_consumption(consumption = 1)  # in £/MWh
```

is this correct? so that people know what the 1 stands for

sofiapinto commented on asf_lifetime_cost_model/analysis/notebooks/data_getters_demo.py:
```diff
@@ -0,0 +1,215 @@
+# If we want to extract a household's total gas bill (assuming we know their gas consumption) and including VAT:
+
+# %%
+current_gas_tariff.calculate_total_consumption(consumption=2.7, vat=True)
```
across what time? month/year?

eglucas replied:
Will add a note!

sofiapinto commented on asf_lifetime_cost_model/analysis/notebooks/data_getters_demo.py:
```diff
@@ -0,0 +1,215 @@
+rebalanced_gas_tariff.calculate_total_consumption(consumption=2.7, vat=True)
+
+# %% [markdown]
+# The gas bill after rebalancing is higher because we moved some policy costs (RO and FiT) from electricity to gas.
```
```suggestion
# In the example above, the gas bill after rebalancing is higher because we moved some policy costs (RO and FiT) from electricity to gas.
```
just so that anyone who reads this knows it's not always de case.

sofiapinto reviewed (commented):
Thanks so much again for this work @eglucas!

I have now done a thorough review of your code and left some additional comments about documentation, as well as questions where I still don't quite understand something. Most of them you should just be able to either accept or reject my suggestions.

Everything looks good from a logic perspective and I've identified some opportunities for improvements, which we can take forward in a following PR.

eglucas committed:
Update docstring asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update docstring asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update docstring asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update docstring asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update function name asf_lifetime_cost_model/getters/getter_utils.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/getters/data_getters.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/analysis/notebooks/data_getters_demo.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
Update asf_lifetime_cost_model/analysis/notebooks/data_getters_demo.py

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

eglucas committed:
update wholesale price projections getter

eglucas committed:
update levies getter

eglucas committed:
update levies rebalancing doc

eglucas committed:
update demo notebook

eglucas commented:
> Thanks so much again for this work @eglucas!
> 
> I have now done a thorough review of your code and left some additional comments about documentation, as well as questions where I still don't quite understand something. Most of them you should just be able to either accept or reject my suggestions.
> 
> Everything looks good from a logic perspective and I've identified some opportunities for improvements, which we can take forward in a following PR.

Thanks for your thorough review and really helpful comments @sofiapinto ! I have actioned all your suggestions, and have made changes to the code and documentation. In addition to committing your suggestions directly, I have:
- removed the projection scenario names argument in the price projections getter, and it now returns all data for all scenarios in the dataframe
- streamlined the code as per your suggestion on the conditionals for including/excluding ncc levy in the levies getter
- added further documentation to explain the meaning of 0 and 1 rebalancing weights in the rebalanced levies getter
- made improvements to the demo notebook (removing redundant pandas import and adding further explanation to consumption units)

sofiapinto commented:
Amazing work ✨ thank you so much! Happy for you to merge!
